From: Miroslav Benes <mbenes@suse.cz>
Date: Thu, 17 Mar 2016 14:04:10 +0100
Subject: kgr: do not return and print an error only if the object is not
 loaded
Patch-mainline: not yet, kgraft
References: fate#313296

When a patch module is loaded an error message is printed to the kernel
log about all unresolved functions provided their objname is set to
NULL. This means vmlinux and such functions should be present all the
time. The error message is not printed for modules at all. However it
should be if the module is loaded.

Signed-off-by: Miroslav Benes <mbenes@suse.cz>
Reviewed-by: Petr Mladek <pmladek@suse.com>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 kernel/kgraft.c |   19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

--- a/kernel/kgraft.c
+++ b/kernel/kgraft.c
@@ -376,6 +376,20 @@ static void kgr_wakeup_kthreads(void)
 	read_unlock(&tasklist_lock);
 }
 
+static bool kgr_is_object_loaded(const char *objname)
+{
+	struct module *mod;
+
+	if (!objname)
+		return true;
+
+	mutex_lock(&module_mutex);
+	mod = find_module(objname);
+	mutex_unlock(&module_mutex);
+
+	return !!mod;
+}
+
 static unsigned long kgr_get_function_address(const struct kgr_patch_fun *pf)
 {
 	unsigned long orig_addr;
@@ -384,7 +398,7 @@ static unsigned long kgr_get_function_ad
 
 	orig_addr = kallsyms_lookup_name(pf->name);
 	if (!orig_addr) {
-		if (!pf->objname)
+		if (kgr_is_object_loaded(pf->objname))
 			pr_err("kgr: function %s not resolved\n", pf->name);
 		return -ENOENT;
 	}
@@ -625,7 +639,8 @@ static int kgr_patch_code(struct kgr_pat
 			return -EINVAL;
 		err = kgr_init_ftrace_ops(patch_fun);
 		if (err) {
-			if (err == -ENOENT && patch_fun->objname) {
+			if (err == -ENOENT &&
+			    !kgr_is_object_loaded(patch_fun->objname)) {
 				patch_fun->state = KGR_PATCH_SKIPPED;
 				return 0;
 			}
